public/app/plugins/panel/canvas/editor/element/utils.ts

```diff
@@ -42,9 +39,10 @@ export const interpolateVariables = (text: string) => {
 };
 
 export const getRequest = (api: APIEditorConfig) => {
-  const requestHeaders: HeadersInit = [];
+  const endpoint = getEndpoint(interpolateVariables(api.endpoint));
+  const url = new URL(endpoint);
 
-  const url = new URL(interpolateVariables(api.endpoint!));
+  const requestHeaders: Record<string, string> = { 'X-Grafana-Action': '1' };
```

Comment:
We need to set this header after the client-specified headers. Right now it would be possible to bypass the checks by setting` X-Grafana-Action` as a header param with an empty value, since that would override this declaration and `action` would be `""` on line 49 in the middleware (one of the downsides of the Go not exposing `Has` on the `Header` API is that we can't easily tell the difference between a header that isn't set and a header set to an empty string).

Reply:
yep, good catch. probably should match the same changes in the Canvas Actions PR:

https://github.com/grafana/grafana/pull/90677/commits/cabaeaa9fdf72ded3d7bf8036949f3e47e064757